internal/cachecontroller/cache_controller.go

```diff
@@ -172,19 +173,29 @@ func (c *InMemoryCacheController) findChangesAndInvalidate(ctx context.Context,
 		return
 	}
 
+	// changes[0] returns the most recent change
 	entry := &storage.ChangelogCacheEntry{
 		LastModified: changes[0].GetTimestamp().AsTime(),
 	}
 
-	lastInvalidationOccurred := c.cache.Get(cacheKey) != nil
+	lastCacheRecord := c.cache.Get(cacheKey)
 
 	// set changelog entry as soon as possible for subsequent cache
 	// lookups have the entry and not have to wait on the existing singleflight group
 	c.cache.Set(cacheKey, entry, c.ttl)
 
-	timestampOfLastInvalidation := time.Now().Add(-c.ttl)
+	timestampOfLastInvalidation := time.Time{}
+	if lastCacheRecord != nil {
+		decodedRecord, ok := lastCacheRecord.(*storage.ChangelogCacheEntry)
+		if ok {
+			// if the change log cache is available and valid, use the last modified
+			// time to have better consistency. Otherwise, the timestampOfLastInvalidation will
+			// be beginning of time which imply the need to perform read changes and invalidate record
```

Comment:
`// be beginning of time which imply the need to perform read changes and invalidate record`

Nitpick: the 1st part of this comment does not make a lot of sense to me. By the time we are in this line, the ReadChanges call has already happened, irrespective of whether there was an entry in the cache

Maybe say `// be beginning of time which imply the need to invalidate one or more records`